File: 0014-longest-common-prefix/0014-longest-common-prefix.py
from typing import List

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        first_word = strs[0]

        for i in range(len(first_word)):
            current_char = first_word[i]

            for word in strs[1:]:
                # i가 word의 길이 이상이면 해당 위치의 문자가 존재하지 않음
                if i >= len(word):
                    return first_word[:i]

                # 같은 위치의 문자가 다르면 공통 접두사는 여기서 끝남
                if word[i] != current_char:
                    return first_word[:i]

        return first_word

File: 0014-longest-common-prefix/0014-longest-common-prefix.c
char* longestCommonPrefix(char** strs, int strsSize) {
    char* first_word = strs[0];

    for (int i = 0; first_word[i] != '\0'; i++) {
        char current_char = first_word[i];

        for (int j = 1; j < strsSize; j++) {
            /*
             * strs[j][i] == '\0' 이면
             * 현재 문자열이 first_word보다 짧아서 더 이상 공통 접두사가 될 수 없음
             * 
             * strs[j][i] != current_char이면
             * 같은 위치의 문자가 다르므로 공통 접두사가 바뀜
             */
            if (strs[j][i] == '\0' || strs[j][i] != current_char) {
                first_word[i] = '\0';
                return first_word;
            }
        }
    }

    return first_word;
}